pkg/protocol/engine/engine.go

```diff
@@ -223,26 +226,34 @@ func (e *Engine) ProcessBlockFromPeer(block *model.Block, source peer.ID) {
 
 // Reset resets the component to a clean state as if it was created at the last commitment.
 func (e *Engine) Reset() {
+	e.accessMutex.Lock()
+	defer e.accessMutex.Unlock()
+
+	// Reset should be performed in the same order as Shutdown.
 	e.BlockRequester.Clear()
-	e.Storage.Reset()
-	e.EvictionState.Reset()
-	e.Filter.Reset()
-	e.CommitmentFilter.Reset()
-	e.BlockCache.Reset()
-	e.BlockDAG.Reset()
-	e.Booker.Reset()
-	e.Ledger.Reset()
-	e.BlockGadget.Reset()
-	e.SlotGadget.Reset()
-	e.Notarization.Reset()
-	e.Attestations.Reset()
-	e.SybilProtection.Reset()
 	e.Scheduler.Reset()
-	e.TipManager.Reset()
 	e.TipSelection.Reset()
-	e.Retainer.Reset()
+	e.TipManager.Reset()
+	e.Attestations.Reset()
 	e.SyncManager.Reset()
+	e.Notarization.Reset()
+	e.SlotGadget.Reset()
+	e.BlockGadget.Reset()
 	e.UpgradeOrchestrator.Reset()
+	e.SybilProtection.Reset()
+	e.Booker.Reset()
+	e.Ledger.Reset()
+	e.CommitmentFilter.Reset()
+	e.BlockDAG.Reset()
+	e.Filter.Reset()
+	e.Retainer.Reset()
+	e.EvictionState.Reset()
+	e.BlockCache.Reset()
+
+	// Waits for all pending tasks to be processed.
+	e.Workers.WaitChildren()
```

Comment:
Why do you only wait for the tasks to be processed **after** resetting? Naively, I'd say we should wait before so that nothing is happening anymore and that we have a clean slate after the reset

Reply:
We figured that the order of events during Reset should match the order in Shutdown, and we just put it in the same place. Do you think we shoul move it in Shutdown method as well?

I think the most important thing is that it's done before we close the storage, so that if there were any pending tasks in the worker pools, they can finish (especially around acceptance/notarization).